internal/pkg/api/handleCheckin.go

```diff
@@ -337,6 +337,9 @@ func (ct *CheckinT) ProcessRequest(zlog zerolog.Logger, w http.ResponseWriter, r
 	actions, ackToken = convertActions(zlog, agent.Id, pendingActions)
 
 	span, ctx := apm.StartSpan(r.Context(), "longPoll", "process")
+	ctx, cancel := context.WithTimeout(ctx, pollDuration)
```

Comment:
Interesting, evaluating what the timeouts and lifetimes of all the requests in here is actually challenging and I'm not sure if they are right.

This context is not obviously tied to the actual network requests, also at this point, we are passed auth, which should probably also respect a timeout and it's not clear that it is tied to the poll duration either.

What this does is cause us to hit the `ctx.Done()` block below, which triggers us to hit the `ct.writeResponse` call. That call is a network operation that should also have a timeout, but it can't be tied to this context because it is expired, so we'd need a different one.

Reply:
Looking closer, this context deadline actually changes the checkin logic without setting a deadline on any of the underlying network operations or interactions with ES.

There is already a ticker for the long poll duration: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/api/handleCheckin.go#L316-L319

It causes us to hit the `CheckIn` method here: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/api/handleCheckin.go#L374-L379

So all setting this context deadline does is get us to this block, but only if we aren't already in the CheckIn method: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/api/handleCheckin.go#L345-L356

If both `<-ctx.Done()` and `<-tick.C` are pending at the same time the Go runtime will randomly choose which case is taken, so the behavior here isn't even deterministic.

Reply:
Thanks for looking into it. Do you think there is a way to enforce the pollDuration on the underlying network requests? Maybe someone from the Control Plane team can spend some time on this, it seems to be the reason of some drones stuck in a failed checkin state.

Reply:
Looking closer, if we are in this loop, the only place where we interact directly with ES are in `processPolicy` and putting the deadline on the context solves that case: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/api/handleCheckin.go#L364

If I follow this all the way down the call stack to where the actual search call happens, the context should prevent us from waiting forever for a response:

https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/bulk/engine.go#L568-L574

The one that is actually on the underlying ES network request is in the bulker run function: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/bulk/engine.go#L356

The context for this appears to just be tied to `context.Background` in multiple places: https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/server/fleet.go#L374-L379

https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/bulk/engine.go#L166-L182

It does look like there may be a default 90s timeout on the underlying ES client, but I don't see this actually being called anywhere (possible I missed it): https://github.com/elastic/fleet-server/blob/7ecbda15c53ba0bcedb9507cc49dc35311513e0e/internal/pkg/config/output.go#L59-L62

There are lots of places that could be the problem, TBH I'd just add more logging or spans until we definitively know exactly where we get stuck when these 28+ minute checkins happen.

Reply:
> Maybe someone from the Control Plane team can spend some time on this, it seems to be the reason of some drones stuck in a failed checkin state.

For now that person is me, let's avoid context switching in someone else while we narrow down what is actually wrong and can evaluate the effort to fix it.